File: leetcode/weekly-contest-176/1352.cc
#include <vector>


using namespace std;


class ProductOfNumbers {
  private:
    vector<int> buf;
  public:
    ProductOfNumbers() {
    }

    void add(int num) {
        buf.emplace_back(num);
    }

    int getProduct(int k) {
        int result = 1;
        for (int i = buf.size()-1; k > 0; --i, --k) {
            result *= buf[i];
            if (!result)
                break;
        }
        return result;
    }
};

/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * ProductOfNumbers* obj = new ProductOfNumbers();
 * obj->add(num);
 * int param_2 = obj->getProduct(k);
 */

File: leetcode/1352.py
class ProductOfNumbers:
    def __init__(self):
        self.buf = [1]

    def add(self, num: int) -> None:
        if num == 0:
            self.buf = [1]
        else:
            self.buf.append(self.buf[-1] * num)

    def getProduct(self, k: int) -> int:
        if k >= len(self.buf):
            return 0
        return self.buf[-1] // self.buf[-k-1]


# Your ProductOfNumbers object will be instantiated and called as such:
# obj = ProductOfNumbers()
# obj.add(num)
# param_2 = obj.getProduct(k)
